### PR TITLE
New Feat: Added HNSWLib and in-memory self-query retriever

### PR DESCRIPTION
**Pull request by ppramesi**
Added HNSWLib and in-memory self-query retriever. Also added docs, examples and integration tests. Also added filter for in-memory vector store.

### CODE REVIEW

File: docs/docs/modules/indexes/retrievers/hnswlib-self-query.mdx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Self Query HNSWLib Retriever
```
Comment (jacoblee93) [inline docs/docs/modules/indexes/retrievers/hnswlib-self-query.mdx:1]: I'll move these into their own section of the retriever docs

**Review by jacoblee93 — COMMENTED** [review]
Looks awesome, one small question

File: langchain/src/retrievers/self_query/functional_translator.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  getComparatorFunction<C extends Comparator>(
+    comparator: Comparator
+  ): (a: string | number, b: ValueType[C]) => boolean {
+    switch (comparator) {
+      case Comparators.eq: {
+        return (a: string | number, b: ValueType[C]) => a === b;
+      }
+      case Comparators.ne: {
```
[the remark this answers is not in the document]
Comment (ppramesi) [reply langchain/src/retrievers/self_query/functional_translator.ts:58]: I messed up here. `ne` and `neq` are shorthand for the same comparator "not equal", and there should be only one of them and it should be `ne`. I'm going to change all occurrences of `neq` and change it to `ne` here and the other 2 pull requests.

File: langchain/src/retrievers/self_query/tests/hnswlib_self_query.int.test.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import { test } from "@jest/globals";
+import { Document } from "../../../document.js";
+import { AttributeInfo } from "../../../schema/query_constructor.js";
+import { OpenAIEmbeddings } from "../../../embeddings/openai.js";
+import { SelfQueryRetriever } from "../index.js";
+import { OpenAI } from "../../../llms/openai.js";
+import { FunctionalTranslator } from "../functional_translator.js";
+import { HNSWLib } from "../../../vectorstores/hnswlib.js";
+
+test.skip("HNSWLib Store Self Query Retriever Test", async () => {
```
Comment (jacoblee93) [inline langchain/src/retrievers/self_query/tests/hnswlib_self_query.int.test.ts:10]: Also @ppramesi this caused OOM errors for me sporadically - any idea why?

**Comment by jacoblee93** [comment]
`langchain/src/chains/query_constructor/tests/query_chain.int.test.ts` is failing - you can run with:
`yarn test:single langchain/src/chains/query_constructor/tests/query_chain.int.test.ts`

**Comment by ppramesi** [comment]
@jacoblee93 Fixed failing query_chain test and HNSWLib test failing sometimes by removing `in` comparator. The LLM will use multiple `or` comparators instead when building queries.

**Pull request merged**